src/xstate/atomWithMachine.ts

```diff
@@ -1,91 +1,131 @@
   EventObject,
-  Interpreter,
+  InternalMachineOptions,
+  InterpreterFrom,
   InterpreterOptions,
-  MachineOptions,
-  State,
-  StateMachine,
-  Typestate,
+  Prop,
+  StateConfig,
+  StateFrom,
 } from 'xstate'
+import type { Atom, Getter, WritableAtom } from 'jotai'
 import { atom } from 'jotai'
-import type { Atom, Getter } from 'jotai'
+
+export interface MachineAtomOptions<TContext, TEvent extends EventObject> {
+  /**
+   * If provided, will be merged with machine's `context`.
+   */
+  context?: Partial<TContext>
+  /**
+   * The state to rehydrate the machine to. The machine will
+   * start at this state instead of its `initialState`.
+   */
+  state?: StateConfig<TContext, TEvent>
+}
+
+type Options<TMachine extends AnyStateMachine> =
+  AreAllImplementationsAssumedToBeProvided<
+    TMachine['__TResolvedTypesMeta']
+  > extends false
+    ? InterpreterOptions &
+        MachineAtomOptions<TMachine['__TContext'], TMachine['__TEvent']> &
+        InternalMachineOptions<
+          TMachine['__TContext'],
+          TMachine['__TEvent'],
+          TMachine['__TResolvedTypesMeta'],
+          true
+        >
+    : InterpreterOptions &
+        MachineAtomOptions<TMachine['__TContext'], TMachine['__TEvent']> &
+        InternalMachineOptions<
+          TMachine['__TContext'],
+          TMachine['__TEvent'],
+          TMachine['__TResolvedTypesMeta']
+        >
+
+type MaybeParam<T> = T extends (v: infer V) => unknown ? V : never
 
 export function atomWithMachine<
-  TContext,
-  TEvent extends EventObject,
-  TTypestate extends Typestate<TContext> = { value: any; context: TContext }
+  TMachine extends AnyStateMachine,
+  TInterpreter = InterpreterFrom<TMachine>
 >(
-  getMachine:
-    | StateMachine<TContext, any, TEvent, TTypestate>
-    | ((get: Getter) => StateMachine<TContext, any, TEvent, TTypestate>),
-  getOptions?:
-    | (Partial<InterpreterOptions> & Partial<MachineOptions<TContext, TEvent>>)
-    | ((
-        get: Getter
-      ) => Partial<InterpreterOptions> &
-        Partial<MachineOptions<TContext, TEvent>>)
-) {
-  type Machine = StateMachine<TContext, any, TEvent, TTypestate>
-  type Service = Interpreter<TContext, any, TEvent, TTypestate>
-  type MachineState = State<TContext, TEvent, any, TTypestate>
-  const cachedMachineAtom = atom<{ machine: Machine; service: Service } | null>(
-    null
-  )
+  getMachine: TMachine | ((get: Getter) => TMachine),
+  getOptions?: Options<TMachine> | ((get: Getter) => Options<TMachine>)
+): WritableAtom<
+  StateFrom<TMachine>,
+  MaybeParam<Prop<TInterpreter, 'send'>>,
+  void
+> {
+  const cachedMachineAtom = atom<{
+    machine: AnyStateMachine
+    service: AnyInterpreter
+  } | null>(null)
   const machineAtom = atom(
     (get) => {
       const cachedMachine = get(cachedMachineAtom)
       if (cachedMachine) {
         return cachedMachine
       }
       let initializing = true
+      // eslint-disable-next-line @typescript-eslint/explicit-function-return-type
```

Comment:
Is this required?

Reply:
well, probably not in your repo realm, mine has stricter rules, that was a copy from mine, do you want me remove those&

Reply:
done